In [1]:
#imports
import pandas as pd
%matplotlib inline 
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns
from pandas.plotting import register_matplotlib_converters
register_matplotlib_converters()
import calendar
import folium
from folium import plugins
from folium.plugins import HeatMap

In [2]:
# Consignpro's user guide defining some columns
# https://www.consignpro.com/download/UserGuide.pdf

In [3]:
df_customers = pd.read_excel('Customers.xlsx', index_col="CustomerID")

In [4]:
df_customers.head(1)

First Name Last Name Address    City State  Postal Code  \
CustomerID                                                           
0              filler    filler  filler  filler   NaN          NaN   

            Work Phone  Home Phone  AccountBalance OpenedDate  ... CustNotes  \
CustomerID                                                     ...             
0                  NaN         NaN             0.0   1/1/1990  ...       NaN   

            Email  CheckAmt  PickupDate  BirthMonth  PickupDate2  PickupDate3  \
CustomerID                                                                      
0             NaN       0.0         NaT         NaN          NaT          NaT   

            UseMRW TotalSales  Locked  
CustomerID                             
0            False       61.2   False  

[1 rows x 31 columns]

In [5]:
df_activity = pd.read_excel('Activity.xlsx', index_col="ID")

In [6]:
df_activity.head()

CustomerID PostingDate PostingTime  TrxAmount TrxCode Salesperson  \
ID                                                                         
16082          20  2017-09-01    16:58:47   0.000004       O      System   
16105           4  2018-05-23    16:11:45   1.879968       O      System   
16106           6  2015-11-03    18:31:15   0.000000       O      System   
16107          38  2018-04-01    15:50:25   0.000008       O      System   
16108          26  2016-09-01    17:58:53   0.000000       O      System   

       ProductID  DraftNumber  Date1  Text1  Num1 Message  ClearedDate  
ID                                                                      
16082          0            0    NaN    NaN     0     NaN          NaN  
16105          0            0    NaN    NaN     0     NaN          NaN  
16106          0            0    NaN    NaN     0     NaN          NaN  
16107          0            0    NaN    NaN     0     NaN          NaN  
16108          0            0    NaN    NaN     0     NaN          NaN

In [7]:
df_invoices = pd.read_excel('Invoices.xlsx', index_col="Invoice ID")

In [8]:
df_invoices.head()

Employee Invoice Date PaymentTerms  GrossDue  Discount  \
Invoice ID                                                              
0                    cp    12/1/2005          NaN       0.0       0.0   
2439                NaN          NaN          NaN       0.0       0.0   
2964                NaN          NaN          NaN       0.0       0.0   
6457                NaN          NaN          NaN       0.0       0.0   
8112        Voided-Noel    8/13/2018       Refund    -605.0       0.0   

            SalesTax  AcctCredit  NetDue  ItemCount  CreditCardCut  ...  \
Invoice ID                                                          ...   
0               0.00           0    0.00          0              0  ...   
2439            0.00           0    0.00          0              0  ...   
2964            0.00           0    0.00          0              0  ...   
6457            0.00           0    0.00          0              0  ...   
8112          -50.67           0 -655.67          1              0  ...   

            BuyersFee  Layaways  TaxExemptNo LayCredit  RegisterID Advertise  \
Invoice ID                                                                     
0                 NaN       NaN          NaN       NaN         1.0       NaN   
2439              NaN       NaN          NaN       NaN         NaN       NaN   
2964              NaN       NaN          NaN       NaN         NaN       NaN   
6457              NaN       NaN          NaN       NaN         NaN       NaN   
8112              NaN       0.0          NaN       0.0         1.0      8073   

            ConsID  GCSales  GCCredits  CustNo  
Invoice ID                                      
0              NaN      NaN        NaN     NaN  
2439           NaN      NaN        NaN     NaN  
2964           NaN      NaN        NaN     NaN  
6457           NaN      NaN        NaN     NaN  
8112           NaN      0.0        0.0     NaN  

[5 rows x 28 columns]

In [9]:
df_inv_det = pd.read_excel('InvDetails.xlsx', index_col="Invoice Detail ID")

In [10]:
df_inv_det.head()

Invoice ID  Product ID  Price Per Unit  TaxAmount  \
Invoice Detail ID                                                      
0                           0           0             0.0        NaN   
16226                    8113       19708            19.8       1.66   
16227                    8113       20519            22.0       1.84   
16228                    8114       20246           247.0      20.69   
16229                    8116       20049             9.9       0.83   

                   ConsignorID  StoreOwned  ItemCost  FedTax  Voided  TaxRef  \
Invoice Detail ID                                                              
0                          NaN       False       NaN     NaN   False     NaN   
16226                   1362.0       False      9.01     0.0   False     NaN   
16227                      1.0        True      0.00     0.0   False     NaN   
16228                   1382.0       False    112.40     0.0   False     NaN   
16229                   1292.0       False      4.51     0.0   False     NaN   

                   FedTaxRef  DiscRef  CardNumber VoidBy  
Invoice Detail ID                                         
0                        NaN      NaN         NaN    NaN  
16226                    NaN      NaN         NaN    NaN  
16227                    NaN      NaN         NaN    NaN  
16228                    NaN      NaN         NaN    NaN  
16229                    NaN      NaN         NaN    NaN

In [11]:
df_employees = pd.read_excel('Employees.xlsx', index_col="Employee ID")

In [12]:
df_employees.head(1)

Password  AdminPrivledge  Social Security Number FirstName  \
Employee ID                                                              
1              noel1            True             999999999.0      Noel   

            Last Name  Title         Address           City State  \
Employee ID                                                         
1                Ladd    NaN  3704 N Western  Oklahoma City    OK   

             Postal Code  ...  StoreID AdminFin AdminVoid  AdminCons  \
Employee ID               ...                                          
1                  73118  ...        1     True      True       True   

             AdminInv  AdminEmp  AdminAct  AdminBal  MakeSale  NotMakeSale  
Employee ID                                                                 
1                True      True      True      True     False        False  

[1 rows x 24 columns]

In [13]:
df_pay_totals = pd.read_excel('PayTotals.xlsx')

In [14]:
df_pay_totals.head()

PayDate     PayTerms  PayTotal  PayCount
0 2015-09-24        Check     42.27         1
1 2015-09-24  Credit Card    243.84         1
2 2015-09-25  Credit Card      8.67         1
3 2015-09-26        Check    486.60         1
4 2015-09-26  Credit Card   1919.32         3

In [15]:
df_merchandise = pd.read_excel('Merchandise.xlsx', index_col="ProductID")

In [16]:
df_merchandise.head()

ProductType ProductColor ProductSize        ProductDescription  \
ProductID                                                                  
-2               None         None        None                      Lamp   
-1               None         None        None  Wrought Iron W. 5 Chairs   
 0             filler       filler      filler                    filler   
 1               None         None        None             Candle Holder   
 2               None         None        None                     Colum   

           OrigPrice   Price  EnterDate TransferDate  TransferPrice  \
ProductID                                                             
-2            147.60  147.60 2018-08-18   2018-08-18         147.60   
-1            124.75  124.75 2018-07-03   2018-07-03         124.75   
 0               NaN    0.00 1999-12-31   2018-04-14          68.00   
 1              2.00    2.00 2020-05-20   2020-05-20           2.00   
 2             49.50   49.50 2020-07-18   2020-07-18          49.50   

          TransferInd  ... ListOnline  CustNo  FixSplit  TaxRate  PhyDate  \
ProductID              ...                                                  
-2               Sold  ...      False     NaN     False      NaN      NaN   
-1               Sold  ...      False     NaN     False      NaN      NaN   
 0               Sold  ...      False     NaN     False      NaN      NaN   
 1               Sold  ...      False     NaN     False      NaN      NaN   
 2               Sold  ...      False     NaN     False      NaN      NaN   

          Auction2 ConCode BuyOut  CardFee  Reminded  
ProductID                                             
-2             NaN     NaN  False      0.0     False  
-1             NaN     NaN  False      0.0     False  
 0             NaN     NaN  False      0.0     False  
 1             NaN     NaN  False      0.0     False  
 2             NaN     NaN  False      0.0     False  

[5 rows x 50 columns]

In [17]:
df_user_lists = pd.read_excel('UserLists.xlsx', index_col="ID")

In [18]:
df_user_lists.head()

ListNumber         Value  ChargePercent  SplitPct  ConPeriod  Markdown  \
ID                                                                           
75           0          Desk              0       NaN        NaN       NaN   
76           0          Sofa              0       NaN        NaN       NaN   
77           0  Dining Table              0       NaN        NaN       NaN   
78           0    Side Table              0       NaN        NaN       NaN   
79           0  Coffee Table              0       NaN        NaN       NaN   

    MDDay1  MDPct1  MDDay2  MDPct2  ...  NoTax  SmartP  LowP  HighP  TiedTwo  \
ID                                  ...                                        
75     NaN     NaN     NaN     NaN  ...  False     NaN   NaN    NaN      NaN   
76     NaN     NaN     NaN     NaN  ...  False     NaN   NaN    NaN      NaN   
77     NaN     NaN     NaN     NaN  ...  False     NaN   NaN    NaN      NaN   
78     NaN     NaN     NaN     NaN  ...  False     NaN   NaN    NaN      NaN   
79     NaN     NaN     NaN     NaN  ...  False     NaN   NaN    NaN      NaN   

    TaxRate  TotSales  NumSales  Reduce  TaxExceed  
ID                                                  
75      NaN  41680.42     132.0     NaN        NaN  
76      NaN  70513.86     169.0     NaN        NaN  
77      NaN  22310.27      42.0     NaN        NaN  
78      NaN   9730.69      70.0     NaN        NaN  
79      NaN  23344.40     113.0     NaN        NaN  

[5 rows x 22 columns]

## Customers Dataset Cleaning

In [19]:
df_customers.shape

(2272, 31)

In [20]:
df_customers.count()

First Name        2272
Last Name         2272
Address           2272
City              2272
State             2271
Postal Code       2271
Work Phone         915
Home Phone        2243
AccountBalance    2272
OpenedDate        2272
ActivityDate      2272
StoreID           2272
ActiveInd         2272
CheckFee          2272
Date1                0
Text1                0
Num1              2272
ConsignCode          0
Street2            151
Return            2272
ActiveItems        465
CustNotes         2243
Email             2193
CheckAmt          2039
PickupDate         708
BirthMonth           2
PickupDate2        171
PickupDate3         73
UseMRW            2272
TotalSales        2272
Locked            2272
dtype: int64

In [21]:
df_customers.dtypes

First Name                object
Last Name                 object
Address                   object
City                      object
State                     object
Postal Code              float64
Work Phone               float64
Home Phone               float64
AccountBalance           float64
OpenedDate                object
ActivityDate              object
StoreID                    int64
ActiveInd                   bool
CheckFee                   int64
Date1                    float64
Text1                    float64
Num1                       int64
ConsignCode              float64
Street2                   object
Return                      bool
ActiveItems              float64
CustNotes                 object
Email                     object
CheckAmt                 float64
PickupDate        datetime64[ns]
BirthMonth               float64
PickupDate2       datetime64[ns]
PickupDate3       datetime64[ns]
UseMRW                      bool
TotalSales               float64
Locked    

In [22]:
df_customers['StoreID'].unique()
# will drop from dataframe

array([1])

In [23]:
df_customers['ActiveInd'].unique()

array([False,  True])

In [24]:
df_customers['CheckFee'].unique()
# will drop

array([   0, 9999])

In [25]:
df_customers['Num1'].unique()
# will drop

array([ 0, 50])

In [26]:
df_customers['Street2'].unique()

array([nan, 'Unit 3', '#141', 'STE 18-137', 'Apt 2607', '#2209', '#418',
       'Edmond', 'Ste. 505', '#1', 'A204', '#118', '#6311', '#103',
       'Apt K', 'Apt. 924', 'Apt. C', 'Apt. A', 'Villa 82', 'Apt. 1138',
       'Apt #1138', 'Apt 425', 'Apt 204', 'Apt 104', 'Unit 61', '207B',
       '#A120', '#112', '#2222', '#G', '#18201', 'Apt. 452', 'Apt 8210',
       'Apt. 14k1', 'Apt 2320', 'Apt. 612', 'Oak Hollow Rd', 'Apt. 132',
       '#24101', '#909', 'Unit 101', '#803', '#340', 'Ste. 200',
       'Apt 248', '#9', 'Apt. 17D', 'Apt 16108', 'Apt 1005', 'Apt 1512',
       '#225A', 'Unit #185', '#22', '#309', 'Apt # 1702', '9-137',
       'Suite #1550', 'Apt 1428', 'Suite 1403', 'Villa 80', 'Apt #4',
       'Apt 418', 'Dale St', '#2117', 'Apt. 19208', '910 S. Main',
       '#2705', 'Suite 112', '#276', 'Unit 150', 'Apt # 155', '#14',
       '#465', 'Apt 5108', 'Apt 222', '#233', 'Villa 27', 'Apt 134',
       'Suite 200', '#169', 'Unit B', '#201', 'Suite 3121', 'Apt 202',
       '#31', 'AP

In [27]:
df_customers['ActiveItems'].unique()

array([ 0., 44.,  1., 14.,  3.,  2.,  7.,  4., 19.,  6., 76., 11., 57.,
        5., 94., 16.,  8., 17., 12.,  9., 25., 41., nan])

In [28]:
df_customers['CustNotes'].unique()

array([nan, 'Store', 'Owes $50 p/u 9-17-15 njl', ...,
       'PU $75 GG 8/12/2020', 'Brought in 8/12/2020 GG', 'Brout in'],
      dtype=object)

In [29]:
df_customers['CheckAmt'].unique()
# will drop

array([ 0., nan])

In [30]:
df_customers['Locked'].unique()
# will drop

array([False])

In [31]:
# deleting columns that will not be relevant to this analysis
# including last names to maintain privacy
df_customers = df_customers.drop(['Last Name','Work Phone','Home Phone','Date1', 'Text1','ConsignCode','BirthMonth','StoreID','CheckFee',
                                  'Num1','Email','ActiveInd','Return','CustNotes','CheckAmt','UseMRW','Locked'], axis=1)
df_customers.head()

First Name                Address           City State  \
CustomerID                                                          
0              filler                 filler         filler   NaN   
1                 K&N        3704 N. Western  Oklahoma City    OK   
2                Alan  1503 Winding Ridge Rd         Edmond    OK   
3               Sarah       8309 NW 146th St  Oklahoma City    OK   
4             Beverly        3802 US HWY 377            Ada    OK   

            Postal Code  AccountBalance OpenedDate ActivityDate Street2  \
CustomerID                                                                
0                   NaN            0.00   1/1/1990    5/18/2018     NaN   
1               73118.0          118.24  9/18/2015    8/12/2020     NaN   
2               73034.0            0.00  9/18/2015     4/5/2016     NaN   
3               73142.0            5.00  9/22/2015     7/2/2017     NaN   
4               74820.0            1.88  9/23/2015    5/22/2018     NaN   

            ActiveItems PickupDate PickupDate2 PickupDate3  TotalSales  
CustomerID                                                              
0                   0.0        NaT         NaT         NaT      61.200  
1                   0.0 2015-07-20  2016-06-05  2019-04-26  768531.800  
2                   0.0 2015-12-17         NaT         NaT     249.500  
3                   0.0 2015-12-21         NaT         NaT    1501.191  
4                   0.0 2015-12-22         NaT         NaT    9924.199

In [32]:
# drop the first row which is just a filler row
df_customers = df_customers.drop([0])
df_customers.head()

First Name                Address           City State  \
CustomerID                                                          
1                 K&N        3704 N. Western  Oklahoma City    OK   
2                Alan  1503 Winding Ridge Rd         Edmond    OK   
3               Sarah       8309 NW 146th St  Oklahoma City    OK   
4             Beverly        3802 US HWY 377            Ada    OK   
5               Juana       702 Wakefield Rd          Perry    OK   

            Postal Code  AccountBalance OpenedDate ActivityDate Street2  \
CustomerID                                                                
1               73118.0          118.24  9/18/2015    8/12/2020     NaN   
2               73034.0            0.00  9/18/2015     4/5/2016     NaN   
3               73142.0            5.00  9/22/2015     7/2/2017     NaN   
4               74820.0            1.88  9/23/2015    5/22/2018     NaN   
5               73077.0            0.00  9/23/2015    6/29/2020     NaN   

            ActiveItems PickupDate PickupDate2 PickupDate3  TotalSales  
CustomerID                                                              
1                   0.0 2015-07-20  2016-06-05  2019-04-26  768531.800  
2                   0.0 2015-12-17         NaT         NaT     249.500  
3                   0.0 2015-12-21         NaT         NaT    1501.191  
4                   0.0 2015-12-22         NaT         NaT    9924.199  
5                   0.0        NaT         NaT         NaT    1533.500

In [33]:
# change 'Postal Code' from a float to a string
df_customers['Postal Code'] = df_customers['Postal Code'].astype(str).str[:-2]
df_customers.head()

First Name                Address           City State Postal Code  \
CustomerID                                                                      
1                 K&N        3704 N. Western  Oklahoma City    OK       73118   
2                Alan  1503 Winding Ridge Rd         Edmond    OK       73034   
3               Sarah       8309 NW 146th St  Oklahoma City    OK       73142   
4             Beverly        3802 US HWY 377            Ada    OK       74820   
5               Juana       702 Wakefield Rd          Perry    OK       73077   

            AccountBalance OpenedDate ActivityDate Street2  ActiveItems  \
CustomerID                                                                
1                   118.24  9/18/2015    8/12/2020     NaN          0.0   
2                     0.00  9/18/2015     4/5/2016     NaN          0.0   
3                     5.00  9/22/2015     7/2/2017     NaN          0.0   
4                     1.88  9/23/2015    5/22/2018     NaN          0.0   
5                     0.00  9/23/2015    6/29/2020     NaN          0.0   

           PickupDate PickupDate2 PickupDate3  TotalSales  
CustomerID                                                 
1          2015-07-20  2016-06-05  2019-04-26  768531.800  
2          2015-12-17         NaT         NaT     249.500  
3          2015-12-21         NaT         NaT    1501.191  
4          2015-12-22         NaT         NaT    9924.199  
5                 NaT         NaT         NaT    1533.500

In [34]:
# check that column is of dtype Object ('O')
df_customers['Postal Code'].dtypes

dtype('O')

In [35]:
# convert column datatypes to proper datetime formats
df_customers['OpenedDate'] = pd.to_datetime(df_customers['OpenedDate'], format='%d:%m:%Y', infer_datetime_format=True)
df_customers['ActivityDate'] = pd.to_datetime(df_customers['ActivityDate'], format='%d:%m:%Y', infer_datetime_format=True)
print(df_customers['OpenedDate'].dtype)
print(df_customers['ActivityDate'].dtype)

datetime64[ns]
datetime64[ns]


In [36]:
df_customers.tail()

First Name             Address           City State  \
CustomerID                                                             
2271                  James       763 Mohawk St       Columbus    OH   
2272        Staci And Brian      1201 Irvine Dr         Edmond    OK   
2273                  Louis        200 N Harvey  Oklahoma City    OK   
2274                  Carol  1729 Coventry Lane  Oklahoma City    00   
2275                   Gary     10407 N May Ave  Oklahoma City    OK   

           Postal Code  AccountBalance OpenedDate ActivityDate Street2  \
CustomerID                                                               
2271             43206             0.0 2020-08-11   2020-08-11     NaN   
2272             73025             0.0 2020-08-11   2020-08-12     NaN   
2273             73102           -75.0 2020-08-12   2020-08-12     606   
2274             73120             0.0 2020-08-12   2020-08-12     NaN   
2275             73120             0.0 2020-08-12   2020-08-12     NaN   

            ActiveItems PickupDate PickupDate2 PickupDate3  TotalSales  
CustomerID                                                              
2271                NaN 2020-11-09         NaT         NaT         0.0  
2272                NaN 2020-11-09         NaT         NaT         0.0  
2273                NaN        NaT         NaT         NaT         0.0  
2274                NaN 2020-11-10         NaT         NaT         0.0  
2275                NaN 2020-11-10         NaT         NaT         0.0

In [37]:
df_customers.shape

(2271, 14)

## Activity Dataset Cleaning

In [38]:
df_activity.head()

CustomerID PostingDate PostingTime  TrxAmount TrxCode Salesperson  \
ID                                                                         
16082          20  2017-09-01    16:58:47   0.000004       O      System   
16105           4  2018-05-23    16:11:45   1.879968       O      System   
16106           6  2015-11-03    18:31:15   0.000000       O      System   
16107          38  2018-04-01    15:50:25   0.000008       O      System   
16108          26  2016-09-01    17:58:53   0.000000       O      System   

       ProductID  DraftNumber  Date1  Text1  Num1 Message  ClearedDate  
ID                                                                      
16082          0            0    NaN    NaN     0     NaN          NaN  
16105          0            0    NaN    NaN     0     NaN          NaN  
16106          0            0    NaN    NaN     0     NaN          NaN  
16107          0            0    NaN    NaN     0     NaN          NaN  
16108          0            0    NaN    NaN     0     NaN          NaN

In [39]:
df_activity.dtypes

CustomerID              int64
PostingDate    datetime64[ns]
PostingTime            object
TrxAmount             float64
TrxCode                object
Salesperson            object
ProductID               int64
DraftNumber             int64
Date1                 float64
Text1                 float64
Num1                    int64
Message                object
ClearedDate           float64
dtype: object

In [40]:
df_activity['DraftNumber'].unique()

array([    0, 11353, 11354, ..., 14153, 14154, 14155])

In [41]:
df_activity['Date1'].unique()
# will drop column

array([nan])

In [42]:
df_activity['Text1'].unique()
# will drop column

array([nan])

In [43]:
df_activity['Num1'].unique()
# will drop column

array([0])

In [44]:
df_activity['Message'].unique()

array([nan, 'Discount Adj NJL', 'P/u Fee', 'Delivery Fee',
       'Balance Adjustment.', 'P/u Fee Refunded', 'Adjustment For 21800',
       '3 Piece Pictures', 'Delivery Fee To Storage', 'Item Was Returned',
       'Manual Pay Out', 'Manual Payout', '23501 Full Payout Per Noel',
       'Reverse P/u Fee', 'Rug We Lost', 'Full Amount For Cushions',
       'Pick Up Fee', 'Pickup Fee', 'Manual Adjustment', 'Rug Adj',
       'Pickup Fee = $450', 'Cust Paid Del Fee/ No Pu Fee',
       'Manual Adjust', 'Correction', 'PU Fee 5/30/19', 'Pu 6/1/19 GG',
       'Pu Fee', 'P/u Fee.', 'Less Mileage Credit', 'Pu Fee 6/4/19 GG',
       '20045', 'Correction NJL', 'PU FEE', 'PU Fee Is Only $200.00',
       'Item 24479 Not Theirs', 'Item 24479 Mirror', 'PICKUP FEE',
       'Manual Adj For Item 26378', 'For P/u Fee 6/12/19', 'Adj For Sale',
       'P/u Fee Refund', 'Man Adj', 'Manual Check', 'P/u Feet.',
       'Pickup Fee X 2', 'Pickup Fee.', 'Refund For Delivery Fee',
       'Pickup Fee 11/29/19', 'For 

In [45]:
df_activity['ClearedDate'].unique()
# will drop column

array([nan])

In [46]:
df_activity['Salesperson'].unique()

array(['System', 'Noel', 'Adriana', 'Unknown', 'Elizabeth', 'Chris',
       'Ken', 'Paul', 'Tate', 'Gina', 'Trey', 'Beth', 'AdminID', 'Kerry',
       'Fred', 'Chad'], dtype=object)

In [47]:
# deleting columns that will not be relevant to this analysis
df_activity = df_activity.drop(['Date1','Text1','Num1', 'ClearedDate'], axis=1)
df_activity.head()

CustomerID PostingDate PostingTime  TrxAmount TrxCode Salesperson  \
ID                                                                         
16082          20  2017-09-01    16:58:47   0.000004       O      System   
16105           4  2018-05-23    16:11:45   1.879968       O      System   
16106           6  2015-11-03    18:31:15   0.000000       O      System   
16107          38  2018-04-01    15:50:25   0.000008       O      System   
16108          26  2016-09-01    17:58:53   0.000000       O      System   

       ProductID  DraftNumber Message  
ID                                     
16082          0            0     NaN  
16105          0            0     NaN  
16106          0            0     NaN  
16107          0            0     NaN  
16108          0            0     NaN

## Invoices Dataset Cleaning

In [48]:
df_invoices.head()

Employee Invoice Date PaymentTerms  GrossDue  Discount  \
Invoice ID                                                              
0                    cp    12/1/2005          NaN       0.0       0.0   
2439                NaN          NaN          NaN       0.0       0.0   
2964                NaN          NaN          NaN       0.0       0.0   
6457                NaN          NaN          NaN       0.0       0.0   
8112        Voided-Noel    8/13/2018       Refund    -605.0       0.0   

            SalesTax  AcctCredit  NetDue  ItemCount  CreditCardCut  ...  \
Invoice ID                                                          ...   
0               0.00           0    0.00          0              0  ...   
2439            0.00           0    0.00          0              0  ...   
2964            0.00           0    0.00          0              0  ...   
6457            0.00           0    0.00          0              0  ...   
8112          -50.67           0 -655.67          1              0  ...   

            BuyersFee  Layaways  TaxExemptNo LayCredit  RegisterID Advertise  \
Invoice ID                                                                     
0                 NaN       NaN          NaN       NaN         1.0       NaN   
2439              NaN       NaN          NaN       NaN         NaN       NaN   
2964              NaN       NaN          NaN       NaN         NaN       NaN   
6457              NaN       NaN          NaN       NaN         NaN       NaN   
8112              NaN       0.0          NaN       0.0         1.0      8073   

            ConsID  GCSales  GCCredits  CustNo  
Invoice ID                                      
0              NaN      NaN        NaN     NaN  
2439           NaN      NaN        NaN     NaN  
2964           NaN      NaN        NaN     NaN  
6457           NaN      NaN        NaN     NaN  
8112           NaN      0.0        0.0     NaN  

[5 rows x 28 columns]

In [49]:
# drop the first four rows which
df_invoices = df_invoices.iloc[4:]
df_invoices.head()

Employee Invoice Date PaymentTerms  GrossDue  Discount  \
Invoice ID                                                                 
8112           Voided-Noel    8/13/2018       Refund   -605.00       0.0   
8113               Adriana    8/13/2018  Credit Card     41.80       0.0   
8114             Elizabeth    8/13/2018  Credit Card    247.00       0.0   
8115        Voided-Adriana    8/13/2018       Refund   -328.25       0.0   
8116               Adriana    8/13/2018         Cash      9.90       0.0   

            SalesTax  AcctCredit  NetDue  ItemCount  CreditCardCut  ...  \
Invoice ID                                                          ...   
8112          -50.67           0 -655.67          1              0  ...   
8113            3.50           0   45.30          2              0  ...   
8114           20.69           0  267.69          1              0  ...   
8115          -27.49           0 -355.74          2              0  ...   
8116            0.83           0   10.73          1              0  ...   

            BuyersFee  Layaways  TaxExemptNo LayCredit  RegisterID  \
Invoice ID                                                           
8112              NaN       0.0          NaN       0.0         1.0   
8113             1.78       0.0          NaN       0.0         1.0   
8114            22.20       0.0          NaN       0.0         1.0   
8115              NaN       0.0          NaN       0.0         1.0   
8116             0.89       0.0          NaN       0.0         1.0   

                  Advertise  ConsID  GCSales  GCCredits  CustNo  
Invoice ID                                                       
8112                   8073     NaN      0.0        0.0     NaN  
8113               Drive-By     NaN      0.0        0.0     NaN  
8114        Repeat Customer     NaN      0.0        0.0     NaN  
8115                   8073     NaN      0.0        0.0     NaN  
8116        Repeat Customer     NaN      0.0        0.0     NaN  

[5 rows x 28 columns]

In [50]:
df_invoices.dtypes

Employee          object
Invoice Date      object
PaymentTerms      object
GrossDue         float64
Discount         float64
SalesTax         float64
AcctCredit         int64
NetDue           float64
ItemCount          int64
CreditCardCut      int64
InvCost          float64
FedSalesTax      float64
Date1            float64
Text1             object
Num1               int64
PayTerms2         object
PayAmt1          float64
PayAmt2          float64
BuyersFee        float64
Layaways         float64
TaxExemptNo       object
LayCredit        float64
RegisterID       float64
Advertise         object
ConsID           float64
GCSales          float64
GCCredits        float64
CustNo           float64
dtype: object

In [51]:
df_invoices['Invoice Date'] = pd.to_datetime(df_invoices['Invoice Date'], format='%d:%m:%Y', infer_datetime_format=True)
df_invoices['Invoice Date'].dtype

dtype('<M8[ns]')

In [52]:
df_invoices.count()

Employee         5982
Invoice Date     5982
PaymentTerms     5982
GrossDue         5985
Discount         5985
SalesTax         5985
AcctCredit       5985
NetDue           5985
ItemCount        5985
CreditCardCut    5985
InvCost          5985
FedSalesTax      5985
Date1               0
Text1              76
Num1             5985
PayTerms2         111
PayAmt1          5921
PayAmt2           118
BuyersFee        5906
Layaways         5982
TaxExemptNo         5
LayCredit        5982
RegisterID       5982
Advertise        5920
ConsID              0
GCSales          5982
GCCredits        5982
CustNo              0
dtype: int64

In [53]:
# drop columns with many missing values
df_invoices = df_invoices.drop(['Date1','Text1','TaxExemptNo','ConsID','CustNo','PayTerms2','PayAmt2'], axis=1)

In [54]:
df_invoices.count()

Employee         5982
Invoice Date     5982
PaymentTerms     5982
GrossDue         5985
Discount         5985
SalesTax         5985
AcctCredit       5985
NetDue           5985
ItemCount        5985
CreditCardCut    5985
InvCost          5985
FedSalesTax      5985
Num1             5985
PayAmt1          5921
BuyersFee        5906
Layaways         5982
LayCredit        5982
RegisterID       5982
Advertise        5920
GCSales          5982
GCCredits        5982
dtype: int64

In [55]:
df_invoices.dtypes

Employee                 object
Invoice Date     datetime64[ns]
PaymentTerms             object
GrossDue                float64
Discount                float64
SalesTax                float64
AcctCredit                int64
NetDue                  float64
ItemCount                 int64
CreditCardCut             int64
InvCost                 float64
FedSalesTax             float64
Num1                      int64
PayAmt1                 float64
BuyersFee               float64
Layaways                float64
LayCredit               float64
RegisterID              float64
Advertise                object
GCSales                 float64
GCCredits               float64
dtype: object

## Invoice Details Dataset Cleaning

In [56]:
df_inv_det.head()

Invoice ID  Product ID  Price Per Unit  TaxAmount  \
Invoice Detail ID                                                      
0                           0           0             0.0        NaN   
16226                    8113       19708            19.8       1.66   
16227                    8113       20519            22.0       1.84   
16228                    8114       20246           247.0      20.69   
16229                    8116       20049             9.9       0.83   

                   ConsignorID  StoreOwned  ItemCost  FedTax  Voided  TaxRef  \
Invoice Detail ID                                                              
0                          NaN       False       NaN     NaN   False     NaN   
16226                   1362.0       False      9.01     0.0   False     NaN   
16227                      1.0        True      0.00     0.0   False     NaN   
16228                   1382.0       False    112.40     0.0   False     NaN   
16229                   1292.0       False      4.51     0.0   False     NaN   

                   FedTaxRef  DiscRef  CardNumber VoidBy  
Invoice Detail ID                                         
0                        NaN      NaN         NaN    NaN  
16226                    NaN      NaN         NaN    NaN  
16227                    NaN      NaN         NaN    NaN  
16228                    NaN      NaN         NaN    NaN  
16229                    NaN      NaN         NaN    NaN

In [57]:
df_inv_det = df_inv_det.drop([0])
df_inv_det.head()

Invoice ID  Product ID  Price Per Unit  TaxAmount  \
Invoice Detail ID                                                      
16226                    8113       19708            19.8       1.66   
16227                    8113       20519            22.0       1.84   
16228                    8114       20246           247.0      20.69   
16229                    8116       20049             9.9       0.83   
16230                    8117       20600           189.0      15.83   

                   ConsignorID  StoreOwned  ItemCost  FedTax  Voided  TaxRef  \
Invoice Detail ID                                                              
16226                   1362.0       False      9.01     0.0   False     NaN   
16227                      1.0        True      0.00     0.0   False     NaN   
16228                   1382.0       False    112.40     0.0   False     NaN   
16229                   1292.0       False      4.51     0.0   False     NaN   
16230                      1.0        True      0.00     0.0   False     NaN   

                   FedTaxRef  DiscRef  CardNumber VoidBy  
Invoice Detail ID                                         
16226                    NaN      NaN         NaN    NaN  
16227                    NaN      NaN         NaN    NaN  
16228                    NaN      NaN         NaN    NaN  
16229                    NaN      NaN         NaN    NaN  
16230                    NaN      NaN         NaN    NaN

In [58]:
df_inv_det.count()

Invoice ID        11554
Product ID        11554
Price Per Unit    11554
TaxAmount         11554
ConsignorID       11554
StoreOwned        11554
ItemCost          11554
FedTax            11554
Voided            11554
TaxRef               94
FedTaxRef            94
DiscRef              94
CardNumber            0
VoidBy               94
dtype: int64

In [59]:
# drop columns with many missing values
df_inv_det = df_inv_det.drop(['TaxRef','FedTaxRef','DiscRef','CardNumber','VoidBy'], axis=1)
df_inv_det.head()

Invoice ID  Product ID  Price Per Unit  TaxAmount  \
Invoice Detail ID                                                      
16226                    8113       19708            19.8       1.66   
16227                    8113       20519            22.0       1.84   
16228                    8114       20246           247.0      20.69   
16229                    8116       20049             9.9       0.83   
16230                    8117       20600           189.0      15.83   

                   ConsignorID  StoreOwned  ItemCost  FedTax  Voided  
Invoice Detail ID                                                     
16226                   1362.0       False      9.01     0.0   False  
16227                      1.0        True      0.00     0.0   False  
16228                   1382.0       False    112.40     0.0   False  
16229                   1292.0       False      4.51     0.0   False  
16230                      1.0        True      0.00     0.0   False

In [60]:
df_inv_det.dtypes

Invoice ID          int64
Product ID          int64
Price Per Unit    float64
TaxAmount         float64
ConsignorID       float64
StoreOwned           bool
ItemCost          float64
FedTax            float64
Voided               bool
dtype: object

## Employees Dataset Cleaning

In [61]:
df_employees.head(1)

Password  AdminPrivledge  Social Security Number FirstName  \
Employee ID                                                              
1              noel1            True             999999999.0      Noel   

            Last Name  Title         Address           City State  \
Employee ID                                                         
1                Ladd    NaN  3704 N Western  Oklahoma City    OK   

             Postal Code  ...  StoreID AdminFin AdminVoid  AdminCons  \
Employee ID               ...                                          
1                  73118  ...        1     True      True       True   

             AdminInv  AdminEmp  AdminAct  AdminBal  MakeSale  NotMakeSale  
Employee ID                                                                 
1                True      True      True      True     False        False  

[1 rows x 24 columns]

In [62]:
df_employees = df_employees[['FirstName']].copy()
df_employees

FirstName
Employee ID           
1                 Noel
3                  Ken
9            Elizabeth
15                Gina
16                Beth
20                Fred
21                Chad
22             Matthew

## Pay Totals Dataset Cleaning

In [63]:
df_pay_totals.head()

PayDate     PayTerms  PayTotal  PayCount
0 2015-09-24        Check     42.27         1
1 2015-09-24  Credit Card    243.84         1
2 2015-09-25  Credit Card      8.67         1
3 2015-09-26        Check    486.60         1
4 2015-09-26  Credit Card   1919.32         3

In [64]:
df_pay_totals.dtypes

PayDate     datetime64[ns]
PayTerms            object
PayTotal           float64
PayCount             int64
dtype: object

In [65]:
df_pay_totals.count()

PayDate     3770
PayTerms    3753
PayTotal    3770
PayCount    3770
dtype: int64

## Merchandise Dataset Cleaning

In [66]:
df_merchandise.head()

ProductType ProductColor ProductSize        ProductDescription  \
ProductID                                                                  
-2               None         None        None                      Lamp   
-1               None         None        None  Wrought Iron W. 5 Chairs   
 0             filler       filler      filler                    filler   
 1               None         None        None             Candle Holder   
 2               None         None        None                     Colum   

           OrigPrice   Price  EnterDate TransferDate  TransferPrice  \
ProductID                                                             
-2            147.60  147.60 2018-08-18   2018-08-18         147.60   
-1            124.75  124.75 2018-07-03   2018-07-03         124.75   
 0               NaN    0.00 1999-12-31   2018-04-14          68.00   
 1              2.00    2.00 2020-05-20   2020-05-20           2.00   
 2             49.50   49.50 2020-07-18   2020-07-18          49.50   

          TransferInd  ... ListOnline  CustNo  FixSplit  TaxRate  PhyDate  \
ProductID              ...                                                  
-2               Sold  ...      False     NaN     False      NaN      NaN   
-1               Sold  ...      False     NaN     False      NaN      NaN   
 0               Sold  ...      False     NaN     False      NaN      NaN   
 1               Sold  ...      False     NaN     False      NaN      NaN   
 2               Sold  ...      False     NaN     False      NaN      NaN   

          Auction2 ConCode BuyOut  CardFee  Reminded  
ProductID                                             
-2             NaN     NaN  False      0.0     False  
-1             NaN     NaN  False      0.0     False  
 0             NaN     NaN  False      0.0     False  
 1             NaN     NaN  False      0.0     False  
 2             NaN     NaN  False      0.0     False  

[5 rows x 50 columns]

In [67]:
df_merchandise.shape

(15308, 50)

In [68]:
df_merchandise.count()

ProductType           15308
ProductColor          15308
ProductSize           15308
ProductDescription    15308
OrigPrice             15307
Price                 15308
EnterDate             15308
TransferDate          10923
TransferPrice         15308
TransferInd           15308
TransferBy            15308
CustomerID            15308
ReqPrice              15308
InvoiceNumber         10906
Return                15308
MarkdownDate1          9976
MarkdownDate2          8497
MarkdownDate3          6976
LabelPrinted          15308
MercCut               15307
DownPymt              15308
DownWho                   0
DownDate                  0
ItemCount             15308
Date1                     0
Text1                   395
Num1                  15308
ExpireDate            15308
LayawayGrp                0
PCName                14925
DontPrintTag          15308
ItemNotes                 0
WholesalePrice        15308
BuyersFee             15308
DownPhone                 0
ResaleNow           

In [69]:
df_merchandise['DownPymt'].unique()
# will drop

array([0])

In [70]:
df_merchandise['Num1'].unique()
# will drop

array([0])

In [71]:
df_merchandise['Taxable'].unique()

array([nan, 'Y', 'N'], dtype=object)

In [72]:
df_merchandise['Taxable'].value_counts()
# will drop, seems unnecessary to keep when almost every item is taxable

Y    14910
N       15
Name: Taxable, dtype: int64

In [73]:
df_merchandise['NeverMD'].unique()

array([nan, 'Y', 'N'], dtype=object)

In [74]:
df_merchandise['NeverMD'].value_counts()
# will drop, unable to find documentation for this column description

N    13159
Y     1766
Name: NeverMD, dtype: int64

In [75]:
df_merchandise['TaxRate'].unique()
# will drop, these values will be worthless

array([nan,  0.])

In [76]:
df_merchandise['CardFee'].unique()
# will drop, these values will be worthless

array([ 0., nan])

In [77]:
# drop columns with many missing values and ones that are unnecessary
df_merchandise = df_merchandise.drop(['DownWho','DownDate','Date1','Text1','LayawayGrp','ItemNotes','Num1',
                                      'DownPhone','PreviousOwner','LabelPrinted','PCName','DontPrintTag',
                                      'Taxable','WholesalePrice','AuctionID','CustNo','PhyDate','Auction2',
                                      'DownPymt','ResaleNow','ListOnline','FixSplit','ConCode','NeverMD',
                                      'BuyOut','TaxRate','Reminded','CardFee'], axis=1)

In [78]:
# drop filler row
df_merchandise = df_merchandise.drop([0])
df_merchandise.head()

ProductType ProductColor ProductSize        ProductDescription  \
ProductID                                                                  
-2               None         None        None                      Lamp   
-1               None         None        None  Wrought Iron W. 5 Chairs   
 1               None         None        None             Candle Holder   
 2               None         None        None                     Colum   
 3               None         None        None             Credenza Lamp   

           OrigPrice   Price  EnterDate TransferDate  TransferPrice  \
ProductID                                                             
-2            147.60  147.60 2018-08-18   2018-08-18         147.60   
-1            124.75  124.75 2018-07-03   2018-07-03         124.75   
 1              2.00    2.00 2020-05-20   2020-05-20           2.00   
 2             49.50   49.50 2020-07-18   2020-07-18          49.50   
 3             71.10   71.10 2018-08-25   2018-08-25          71.10   

          TransferInd  ... ReqPrice  InvoiceNumber  Return  MarkdownDate1  \
ProductID              ...                                                  
-2               Sold  ...      0.0         8150.0   False            NaT   
-1               Sold  ...      0.0         7749.0   False            NaT   
 1               Sold  ...      0.0        13211.0   False            NaT   
 2               Sold  ...      0.0        13842.0   False            NaT   
 3               Sold  ...      0.0         8225.0   False            NaT   

           MarkdownDate2 MarkdownDate3 MercCut ItemCount  ExpireDate  \
ProductID                                                              
-2                   NaT           NaT     0.0         0  2018-11-16   
-1                   NaT           NaT     0.5         0  2018-10-01   
 1                   NaT           NaT     0.0         0  2020-08-18   
 2                   NaT           NaT     0.5         0  2020-10-16   
 3                   NaT           NaT     0.0         0  2018-11-23   

           BuyersFee  
ProductID             
-2               0.0  
-1               0.0  
 1               0.0  
 2               0.0  
 3               0.0  

[5 rows x 22 columns]

In [79]:
df_merchandise.dtypes

ProductType                   object
ProductColor                  object
ProductSize                   object
ProductDescription            object
OrigPrice                    float64
Price                        float64
EnterDate             datetime64[ns]
TransferDate          datetime64[ns]
TransferPrice                float64
TransferInd                   object
TransferBy                    object
CustomerID                     int64
ReqPrice                     float64
InvoiceNumber                float64
Return                          bool
MarkdownDate1         datetime64[ns]
MarkdownDate2         datetime64[ns]
MarkdownDate3         datetime64[ns]
MercCut                      float64
ItemCount                      int64
ExpireDate            datetime64[ns]
BuyersFee                    float64
dtype: object

In [80]:
df_merchandise['ProductType'].value_counts()

Accessory          5266
Chair              1194
Decor               655
Table               402
None                384
Pillow              371
Art                 365
Sofa                296
Rug                 293
End Table           228
Lamp                224
Picture             221
Desk                171
Light               155
Bed                 140
Barstool            124
Mirror              113
Coffee Table        111
Dining Table        104
Chest               104
Console              99
Ottoman              93
Cabinet              90
Book                 86
Sidetable            79
Chandelier           77
Bench                65
Nightstand           63
Dresser              60
Clock                54
                   ... 
Sofa And Ottoma       1
Half Moon Table       1
Desk And Hutch        1
Daybed With Tru       1
White Entertain       1
Queen Lowpro Bo       1
Gourd                 1
Bunkbed               1
Stereo Speaker        1
Plasma Tv Cabin       1
Side Tables     

In [81]:
products = df_merchandise['ProductType'].value_counts()
products[products < 10].shape
# there are 1103 entries where the total count for a given Product Type appears less than 10 times. 
# I will likely exclude these entries from the dataframe when analyzing products

(1103,)

In [82]:
df_merchandise['ProductColor'].value_counts()

None        14779
White          53
Brown          44
Black          29
Ivory          25
Flint          22
Driftwoo       19
Silver         17
Linen          13
Sapphire       12
Seaglass       12
Wheat          12
Blue           11
Gold           11
Red            10
Teal            9
Wood            9
Gray            9
Multi           8
Navy            8
Oatmeal         8
Metal           7
Misc            7
Tommy           6
Clear           6
Green           5
Yellow          5
Small           4
Bronze          4
Orange          3
            ...  
Birds           1
71.50           1
Walnut          1
Blond           1
Gry/oran        1
Gallery         1
2 Leaves        1
250.00          1
Rainbow         1
Drk Gree        1
Blk Red         1
Palamino        1
Gld Red         1
Red/gold        1
Blk/Blnd        1
1000.           1
Lime            1
Local           1
Chair           1
Bernhard        1
Gin Patc        1
Blu/brn         1
9x12            1
Gray/pur        1
499       

In [83]:
products_color = df_merchandise['ProductColor'].value_counts()
products_color[products_color < 14779].shape

(133,)

Since there's only 133 entires where the Product Color is any description other than "None", I will remove this column from the dataset as it will not be useful for analysis. 

In [84]:
df_merchandise = df_merchandise.drop(['ProductColor'], axis=1)

In [85]:
df_merchandise['ProductSize'].value_counts()

None          14798
Jubilee          94
Park Avenu       39
Queen            18
Small            18
Medium           17
King             14
Large            13
Pomeroy          11
53006225         10
CM1700            8
B446              8
CM4550            8
CM3500            8
53005996          7
CM3200            7
9ft X 4ft         6
Wood              6
22x22             5
53005413el        3
Round             3
2599-t            2
Chair             2
5X8               2
V750560           2
30"               2
51010881          2
30x30             2
8x10              2
4 Pcs             2
              ...  
Cm1700            1
48x36             1
3x 17             1
Single            1
10 X 13           1
Clkc125           1
2140002           1
72 Inch           1
60                1
Urban Road        1
2 Drawer          1
8X10              1
51003645el        1
Steve Call        1
51003635el        1
Twin              1
ARTAFE2165        1
3130018           1
Oak               1


In [86]:
products_size = df_merchandise['ProductSize'].value_counts()
products_size[products_size < 14798].shape

(200,)

I will again drop this column as only 200 of the rows contain useful information regarding Product Size

In [87]:
df_merchandise = df_merchandise.drop(['ProductSize'], axis=1)

In [88]:
df_merchandise.head()

ProductType        ProductDescription  OrigPrice   Price  EnterDate  \
ProductID                                                                       
-2               None                      Lamp     147.60  147.60 2018-08-18   
-1               None  Wrought Iron W. 5 Chairs     124.75  124.75 2018-07-03   
 1               None             Candle Holder       2.00    2.00 2020-05-20   
 2               None                     Colum      49.50   49.50 2020-07-18   
 3               None             Credenza Lamp      71.10   71.10 2018-08-25   

          TransferDate  TransferPrice TransferInd TransferBy  CustomerID  \
ProductID                                                                  
-2          2018-08-18         147.60        Sold      Chris           1   
-1          2018-07-03         124.75        Sold    Adriana         237   
 1          2020-05-20           2.00        Sold       Gina           1   
 2          2020-07-18          49.50        Sold       Noel         237   
 3          2018-08-25          71.10        Sold    Adriana           1   

           ReqPrice  InvoiceNumber  Return MarkdownDate1 MarkdownDate2  \
ProductID                                                                
-2              0.0         8150.0   False           NaT           NaT   
-1              0.0         7749.0   False           NaT           NaT   
 1              0.0        13211.0   False           NaT           NaT   
 2              0.0        13842.0   False           NaT           NaT   
 3              0.0         8225.0   False           NaT           NaT   

          MarkdownDate3  MercCut  ItemCount ExpireDate  BuyersFee  
ProductID                                                          
-2                  NaT      0.0          0 2018-11-16        0.0  
-1                  NaT      0.5          0 2018-10-01        0.0  
 1                  NaT      0.0          0 2020-08-18        0.0  
 2                  NaT      0.5          0 2020-10-16        0.0  
 3                  NaT      0.0          0 2018-11-23        0.0

In [89]:
df_merchandise['ProductDescription'].value_counts()

Pillow                       611
Lamp                         436
Picture                      347
Art                          315
Chair                        296
Rug                          228
Mirror                       216
Vase                         203
Decor                        202
Chandelier                   145
Barstool                     108
Free                          98
Sofa                          95
Book                          88
Desk                          87
Nightstand                    78
Sidetable                     69
Plate                         68
Office Chair                  65
Finial                        65
Coffeetable                   61
Table                         59
Frame                         56
Ottoman                       53
Pewter Figurine               53
Bookcase                      52
Light Fixture                 51
Painting                      50
Console                       49
Bowl                          48
          

want to add product descriptions to product type field if product type field has "none" for the entry and the descrtipion has something relevant to replace with (i.e. sofa, chair, table, rug, etc.).

## User Lists Dataset Cleaning

In [90]:
df_user_lists.head()

ListNumber         Value  ChargePercent  SplitPct  ConPeriod  Markdown  \
ID                                                                           
75           0          Desk              0       NaN        NaN       NaN   
76           0          Sofa              0       NaN        NaN       NaN   
77           0  Dining Table              0       NaN        NaN       NaN   
78           0    Side Table              0       NaN        NaN       NaN   
79           0  Coffee Table              0       NaN        NaN       NaN   

    MDDay1  MDPct1  MDDay2  MDPct2  ...  NoTax  SmartP  LowP  HighP  TiedTwo  \
ID                                  ...                                        
75     NaN     NaN     NaN     NaN  ...  False     NaN   NaN    NaN      NaN   
76     NaN     NaN     NaN     NaN  ...  False     NaN   NaN    NaN      NaN   
77     NaN     NaN     NaN     NaN  ...  False     NaN   NaN    NaN      NaN   
78     NaN     NaN     NaN     NaN  ...  False     NaN   NaN    NaN      NaN   
79     NaN     NaN     NaN     NaN  ...  False     NaN   NaN    NaN      NaN   

    TaxRate  TotSales  NumSales  Reduce  TaxExceed  
ID                                                  
75      NaN  41680.42     132.0     NaN        NaN  
76      NaN  70513.86     169.0     NaN        NaN  
77      NaN  22310.27      42.0     NaN        NaN  
78      NaN   9730.69      70.0     NaN        NaN  
79      NaN  23344.40     113.0     NaN        NaN  

[5 rows x 22 columns]

In [91]:
df_user_lists.count()

ListNumber       54
Value            54
ChargePercent    54
SplitPct          0
ConPeriod         0
Markdown          0
MDDay1            0
MDPct1            0
MDDay2            0
MDPct2            0
MDDay3            0
MDPct3            0
NoTax            54
SmartP            0
LowP              0
HighP             0
TiedTwo           0
TaxRate           0
TotSales         33
NumSales         33
Reduce            0
TaxExceed         0
dtype: int64

In [92]:
# drop all columns containing no entries
df_user_lists = df_user_lists[['ListNumber','Value','NoTax','TotSales','NumSales']]
df_user_lists.head()

ListNumber         Value  NoTax  TotSales  NumSales
ID                                                     
75           0          Desk  False  41680.42     132.0
76           0          Sofa  False  70513.86     169.0
77           0  Dining Table  False  22310.27      42.0
78           0    Side Table  False   9730.69      70.0
79           0  Coffee Table  False  23344.40     113.0

In [93]:
df_user_lists['NoTax'].value_counts()
# will drop

False    54
Name: NoTax, dtype: int64

In [94]:
df_user_lists.shape

(54, 5)

In [95]:
df_user_lists['ListNumber'].value_counts()
# doesn't seem to be of any value for analysis, drop

0    14
6    12
1     8
5     7
3     6
2     4
4     3
Name: ListNumber, dtype: int64

In [96]:
df_user_lists = df_user_lists.drop(['NoTax', 'ListNumber'], axis=1)
df_user_lists.head()

Value  TotSales  NumSales
ID                                  
75          Desk  41680.42     132.0
76          Sofa  70513.86     169.0
77  Dining Table  22310.27      42.0
78    Side Table   9730.69      70.0
79  Coffee Table  23344.40     113.0

In [97]:
df_user_lists.dtypes

Value        object
TotSales    float64
NumSales    float64
dtype: object

In [98]:
df_user_lists['NumSales'] = df_user_lists['NumSales'].fillna(0).astype(int)

In [99]:
df_user_lists.dtypes

Value        object
TotSales    float64
NumSales      int64
dtype: object

In [100]:
df_user_lists.head()

Value  TotSales  NumSales
ID                                  
75          Desk  41680.42       132
76          Sofa  70513.86       169
77  Dining Table  22310.27        42
78    Side Table   9730.69        70
79  Coffee Table  23344.40       113

# Analysis: Where do customers come from?
## Look at
### - Map all customers
### - Bin by total sales and view highest paying customers
### - Identify recurring customers and locations
### - Identify customer areas in relation to competitors
### - 

In [101]:
df_customers.head()

First Name                Address           City State Postal Code  \
CustomerID                                                                      
1                 K&N        3704 N. Western  Oklahoma City    OK       73118   
2                Alan  1503 Winding Ridge Rd         Edmond    OK       73034   
3               Sarah       8309 NW 146th St  Oklahoma City    OK       73142   
4             Beverly        3802 US HWY 377            Ada    OK       74820   
5               Juana       702 Wakefield Rd          Perry    OK       73077   

            AccountBalance OpenedDate ActivityDate Street2  ActiveItems  \
CustomerID                                                                
1                   118.24 2015-09-18   2020-08-12     NaN          0.0   
2                     0.00 2015-09-18   2016-04-05     NaN          0.0   
3                     5.00 2015-09-22   2017-07-02     NaN          0.0   
4                     1.88 2015-09-23   2018-05-22     NaN          0.0   
5                     0.00 2015-09-23   2020-06-29     NaN          0.0   

           PickupDate PickupDate2 PickupDate3  TotalSales  
CustomerID                                                 
1          2015-07-20  2016-06-05  2019-04-26  768531.800  
2          2015-12-17         NaT         NaT     249.500  
3          2015-12-21         NaT         NaT    1501.191  
4          2015-12-22         NaT         NaT    9924.199  
5                 NaT         NaT         NaT    1533.500

In [102]:
# Display top 10 area codes customers come from
df_customers['Postal Code'].value_counts().head(10)

73120    205
73013    198
73118    179
73012    141
73116    119
73034    101
73112     86
73142     85
73162     74
73099     71
Name: Postal Code, dtype: int64

In [103]:
df_customers.isnull()['Address'].value_counts()

False    2271
Name: Address, dtype: int64

In [104]:
df_customers.isnull()['City'].value_counts()

False    2271
Name: City, dtype: int64

In [105]:
df_customers.isnull()['State'].value_counts()

False    2271
Name: State, dtype: int64

In [106]:
df_customers.isnull()['Postal Code'].value_counts()

False    2271
Name: Postal Code, dtype: int64

In [107]:
df_customers_address = df_customers[['Address','City','State','Postal Code']]

In [108]:
df_customers_address.head()

Address           City State Postal Code
CustomerID                                                        
1                 3704 N. Western  Oklahoma City    OK       73118
2           1503 Winding Ridge Rd         Edmond    OK       73034
3                8309 NW 146th St  Oklahoma City    OK       73142
4                 3802 US HWY 377            Ada    OK       74820
5                702 Wakefield Rd          Perry    OK       73077

In [109]:
df_customers_address.shape

(2271, 4)

In [110]:
df_customers_address['Full Address'] = df_customers_address['Address'] + ", " + df_customers_address['City'] + ", " + df_customers_address['State'] + " " + df_customers_address['Postal Code']

/Users/zachcornelison/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [111]:
df_customers_address.head()

Address           City State Postal Code  \
CustomerID                                                           
1                 3704 N. Western  Oklahoma City    OK       73118   
2           1503 Winding Ridge Rd         Edmond    OK       73034   
3                8309 NW 146th St  Oklahoma City    OK       73142   
4                 3802 US HWY 377            Ada    OK       74820   
5                702 Wakefield Rd          Perry    OK       73077   

                                         Full Address  
CustomerID                                             
1            3704 N. Western, Oklahoma City, OK 73118  
2             1503 Winding Ridge Rd, Edmond, OK 73034  
3           8309 NW 146th St, Oklahoma City, OK 73142  
4                      3802 US HWY 377, Ada, OK 74820  
5                   702 Wakefield Rd, Perry, OK 73077

In [112]:
address_list = df_customers_address[['Full Address']]
address_list.head()

Full Address
CustomerID                                           
1            3704 N. Western, Oklahoma City, OK 73118
2             1503 Winding Ridge Rd, Edmond, OK 73034
3           8309 NW 146th St, Oklahoma City, OK 73142
4                      3802 US HWY 377, Ada, OK 74820
5                   702 Wakefield Rd, Perry, OK 73077

In [113]:
# export to csv to input to online reverse geocoder to obtain coordinates to merge
#address_list.to_csv('addresses.csv', header=True)

In [114]:
# Working code trying to use geocodio API

#my_list = df_customers_address['Full Address'].to_list()
#from geocodio import GeocodioClient
#client = GeocodioClient('myAPIkey')
#locations = client.geocode(my_list)
#locations

In [115]:
# Working code trying to use geocodio API

#import json
#from pandas.io.json import json_normalize   
#addys = json_normalize(locations, record_path=['results'])
#addys = addys[['location']]
#addys.head(3)

In [116]:
# Working code trying to use geocodio API

#addys['location'] = addys['location'].astype(str)
#addys = addys.location.str.strip('{}').str.split(', ', expand=True).rename(columns={0:'Latitude', 1:'Longitude'})
#addys['Latitude'] = addys['Latitude'].str[7:]
#addys['Longitude'] = addys['Longitude'].str[7:].map(lambda x: x.rstrip('}"'))
#addys['Longitude'] = addys['Longitude'].map(lambda x: x.rstrip('}"'))
#addys.head(3)

In [117]:
# Import new csv file containing proper coordinates ran through geocoder
missing_addresses = pd.read_csv('missing_addresses.csv', index_col='CustomerID')
missing_addresses.head(3)

Full Address   Latitude  Longitude
CustomerID                                                                 
1            3704 N. Western, Oklahoma City, OK 73118  35.507996 -97.529520
2             1503 Winding Ridge Rd, Edmond, OK 73034  35.670379 -97.394095
3           8309 NW 146th St, Oklahoma City, OK 73142  35.620098 -97.662728

In [118]:
# Full outer join to merge location data with coordinates on CustomerID key value
df_addresses = pd.merge(df_customers_address, missing_addresses, on='CustomerID', how='outer')

In [119]:
df_addresses.head(3)

Address           City State Postal Code  \
CustomerID                                                           
1                 3704 N. Western  Oklahoma City    OK       73118   
2           1503 Winding Ridge Rd         Edmond    OK       73034   
3                8309 NW 146th St  Oklahoma City    OK       73142   

                                       Full Address_x  \
CustomerID                                              
1            3704 N. Western, Oklahoma City, OK 73118   
2             1503 Winding Ridge Rd, Edmond, OK 73034   
3           8309 NW 146th St, Oklahoma City, OK 73142   

                                       Full Address_y   Latitude  Longitude  
CustomerID                                                                   
1            3704 N. Western, Oklahoma City, OK 73118  35.507996 -97.529520  
2             1503 Winding Ridge Rd, Edmond, OK 73034  35.670379 -97.394095  
3           8309 NW 146th St, Oklahoma City, OK 73142  35.620098 -97.662728

In [120]:
# drop second full address column and rename to remove suffix
df_addresses = df_addresses.drop(['Full Address_y'], axis=1)
df_addresses = df_addresses.rename(columns={"Full Address_x": "Full Address"})
df_addresses.head(3)

Address           City State Postal Code  \
CustomerID                                                           
1                 3704 N. Western  Oklahoma City    OK       73118   
2           1503 Winding Ridge Rd         Edmond    OK       73034   
3                8309 NW 146th St  Oklahoma City    OK       73142   

                                         Full Address   Latitude  Longitude  
CustomerID                                                                   
1            3704 N. Western, Oklahoma City, OK 73118  35.507996 -97.529520  
2             1503 Winding Ridge Rd, Edmond, OK 73034  35.670379 -97.394095  
3           8309 NW 146th St, Oklahoma City, OK 73142  35.620098 -97.662728

In [121]:
# ensure shape matches up (result is 1 less for address which coordinates could not be derived)
df_addresses.shape

(2271, 7)

In [122]:
df_addresses.dtypes

Address          object
City             object
State            object
Postal Code      object
Full Address     object
Latitude        float64
Longitude       float64
dtype: object

In [123]:
#generate a folium map to view locations of customers
def generateBaseMap(default_location=[35.507996, -97.529520], default_zoom_start=11):
    base_map = folium.Map(location=default_location, control_scale=True, zoom_start=default_zoom_start)
    return base_map

In [124]:
customers = df_addresses.dropna()

In [125]:
#create an instance of the map
basemap = generateBaseMap()

In [126]:
# render map
customers.apply(lambda row: folium.Circle(location=[row['Latitude'], row['Longitude']],
                                    radius=100, color='Blue', fill_color='Blue').add_to(basemap), axis=1)

basemap